bridge/client/app/_interceptors/http-error-interceptor.ts

```diff
@@ -1,23 +1,37 @@
-import { Injectable } from '@angular/core';
-import {HttpErrorResponse, HttpEvent, HttpHandler, HttpInterceptor, HttpRequest} from "@angular/common/http";
-import {Observable, throwError} from "rxjs";
-import {catchError, retryWhen} from "rxjs/operators";
-import {genericRetryStrategy} from "./http-generic-retry-strategy";
-import {DtToast} from "@dynatrace/barista-components/toast";
+import {Injectable} from '@angular/core';
+import {HttpErrorResponse, HttpEvent, HttpHandler, HttpInterceptor, HttpRequest} from '@angular/common/http';
+import {Observable, of, throwError} from 'rxjs';
+import {catchError, retryWhen} from 'rxjs/operators';
+import {genericRetryStrategy} from './http-generic-retry-strategy';
+import {DtToast} from '@dynatrace/barista-components/toast';
+import {Router} from '@angular/router';
 
 @Injectable({
   providedIn: 'root'
 })
 export class HttpErrorInterceptor implements HttpInterceptor {
+  private isReloading = false;
 
-  constructor(private readonly toast: DtToast){
+  constructor(private readonly toast: DtToast,
+              private readonly router: Router) {
   }
 
   intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
     return next.handle(request)
       .pipe(
         retryWhen(genericRetryStrategy()),
         catchError((error: HttpErrorResponse) => {
+
+          if (error.status === 401) {
+            if (!this.isReloading) {
+              this.isReloading = true;
+              this.toast.create('Login required. Redirecting to login.');
+              setTimeout(() => window.location.href = '/login', 1000);
```

Comment:
Also here.

Neither `setTimeout` nor `window.location.href` is something that I would expect in an Angular app.

Reply:
Here we have the same as with the `/logout` route, also the `/login` route is not handled by the Angular app, but for login we only redirect to the SSO handler, which provides the login mask. The timeout is there, to allow the user to read the Toast with the message `'Login required. Redirecting to login.'` and I think setTimeout is still the best solution therefore.

Reply:
Thanks for the explanation. I would have expected rxjs timeout here, instead of `setTimeout`. Anyway, let's add a comment here for setTimeout, e.g.:
```
// wait for 1 second such that the user can see the toast notification
```

Reply:
@ermin-muratovic 's explanation is correct.

When approaching the solution, I had two options to implement the service interactions and routing

1. Implement routing in Angular
2. Implement routing in Express

I went ahead with the second option. Justification is that this approach reduced the complexity of the solution. 

If I went ahead with Angular, then I would have to introduce another API layer, which works independent of session cookie. And the login would need to be implemented in three layers - Angular : Express : OAuth Service  (For example, authorization request would need to be obtained by Angular application from Express layer. To serve this, express layer interacts with OAuth service. Same goes for redirect handling).

Regardless,  window.location.href is the desired way to navigate to a non angular route. And since navigation destroy the context, I had to delay the navigation so we can show a toast and make user aware what happens.